src/main/java/bio/terra/cda/app/service/QueryService.java

```diff
@@ -57,45 +66,90 @@ private Job runJob(Job queryJob) {
+      // Each node is a single row in the result.
+      for (JsonNode jsonNode : jsonData) {
+        List<String> systems = jsonNode.findValuesAsText("system");
+        Arrays.stream(Source.values())
+            .forEach(
+                s -> {
+                  // Each row can match more than one system, so we have to count them all.
+                  if (systems.contains(s.name())) {
+                    resultsCount.put(s, resultsCount.getOrDefault(s, 0) + 1);
+                  }
+                });
+      }
+    }
+  }
+
+  /**
+   * Convert a BQ value to a json node.
+   *
+   * @param value the value to convert
+   * @param field the schema field for the value
+   * @return a json node for the value
+   */
+  private JsonNode valueToJson(FieldValue value, Field field) {
+    switch (value.getAttribute()) {
+      case RECORD:
+        var object = objectMapper.createObjectNode();
+        var list = value.getRecordValue();
+        var subFields = field.getSubFields();
+        for (int i = 0; i < subFields.size(); i++) {
+          var subField = subFields.get(i);
+          object.set(subField.getName(), valueToJson(list.get(i), subField));
+        }
+        return object;
+      case REPEATED:
+        var array = objectMapper.createArrayNode();
+        for (FieldValue fieldValue : value.getRepeatedValue()) {
+          array.add(valueToJson(fieldValue, field));
+        }
+        return array;
+      case PRIMITIVE:
+        if (value.isNull()) {
+          return NullNode.instance;
+        }
+        switch (field.getType().getStandardType()) {
+          case NUMERIC:
+            return new DecimalNode(value.getNumericValue());
+          case BOOL:
+            return BooleanNode.valueOf(value.getBooleanValue());
+          default:
+            // Primitive types other than boolean and numeric are represented as strings.
+            return new TextNode(value.getStringValue());
+        }
+      default:
+        throw new RuntimeException("Unknown field value type: " + value.getAttribute());
+    }
   }
 
   private Results getJobResults(Job queryJob) {
     try {
       // Get the results.
       TableResult result = queryJob.getQueryResults();
+      FieldList fields = result.getSchema().getFields();
 
       Results results = new Results();
 
-      // Copy all row data to results. Each row is a JSON object.
+      // Copy all row data to results. For each row, create a Json object using the result's schema.
       for (FieldValueList row : result.iterateAll()) {
-        var rowData = row.get(0).getStringValue();
-        results.jsonData.add(rowData);
-        Arrays.stream(Source.values())
-            .forEach(
-                s -> {
-                  // Each row can match one or more system.
-                  if (s.match(rowData)) {
-                    results.resultsCount.put(s, results.resultsCount.getOrDefault(s, 0) + 1);
-                  }
-                });
+        results.jsonData.add(
```

Comment:
I'm sort of surprised...did `objectMapper.readValue(FieldValue.of(FieldValue.Attribute.RECORD, row), JsonNode.class)` not do the trick?

Reply:
Not sure how that would work. `readValue()` is used to convert text to a JSON object, and but what we have here is a bigquery object, we don't have the source JSON that bigquery used to create the table result.

Reply:
Shoot, sorry, I was doing it backwards.  I meant `objectMapper.writeObjectAsString(FieldValue.of(FieldValue.Attribute.RECORD, row))`. That would spit out the actual json string, which is a little different but might simplify the overall approach

Reply:
ObjectMapper doesn't know how to convert a google field value to a json string, I get the error:
```Method threw 'com.fasterxml.jackson.databind.JsonMappingException' exception.```
 It's possible there's a way to intercept the raw json data coming back from the API, I can take another look at that.